engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/MergedDataIndex.java

```diff
@@ -113,65 +140,185 @@ public Map<ColumnSource<?>, String> keyColumnNamesByIndexedColumn() {
+            this.source = source;
+            this.results = new AtomicReferenceArray<>(capacity);
+        }
+
+        RowSet get(final long rowKey) {
+            if (rowKey < 0 || rowKey >= results.length()) {
+                return null;
+            }
+
+            final int iRowKey = (int) rowKey;
+            do {
+                final Object localResult = results.get(iRowKey);
+                if (localResult instanceof RowSet) {
+                    return (RowSet) localResult;
+                }
+                if (localResult instanceof Exception) {
+                    throw new UncheckedDeephavenException("Exception found for cached RowSet",
+                            (Exception) localResult);
+                }
+
+                if (localResult != null) {
+                    // noinspection EmptySynchronizedStatement
+                    synchronized (localResult) {
+                        // don't care to do anything, we are just waiting for the barrier to be done
+                    }
+                    continue;
+                }
+
+                // we need to create our own placeholder, and synchronize on it first
+                final ReentrantLock placeholder = new ReentrantLock();
+                // noinspection SynchronizationOnLocalVariableOrMethodParameter
+                synchronized (placeholder) {
+                    if (!results.compareAndSet(iRowKey, null, placeholder)) {
+                        // we must try again, someone else has claimed the placeholder
+                        continue;
+                    }
+                    // it is our responsibility to get the right answer
+                    final ObjectVector<RowSet> inputRowsets = source.get(rowKey);
+                    Assert.neqNull(inputRowsets, "inputRowsets");
+                    assert inputRowsets != null;
+
+                    // need to get the value and set it into our own value
+                    final RowSet computedResult;
+                    try {
+                        computedResult = mergeRowSets(rowKey, inputRowsets);
+                    } catch (Exception e) {
+                        if (!results.compareAndSet(iRowKey, placeholder, e)) {
+                            throw new IllegalStateException("another thread changed our cache placeholder!");
+                        }
+                        throw e;
+                    }
+                    if (!results.compareAndSet(iRowKey, placeholder, computedResult)) {
+                        throw new IllegalStateException("another thread changed our cache placeholder!");
+                    }
+
+                    return computedResult;
+                }
+            } while (true);
+        }
+    }
+
+    private Table buildTable(final boolean lazyRowsetMerge) {
+        if (lazyTable != null) {
+            if (lazyRowsetMerge) {
+                return lazyTable;
+            } else {
+                indexTable = lazyTable.select();
```

Comment:
This might be worse for cases where you've only loaded a few row sets out of many.

Reply:
I have changed the logic to keep the PartitionedTable around so that the key column select can be reused.  It would be nice to actually be able to reuse the grouping; but that seems less practical.